Context exception with cortex-a57 CPU

### Description

I face a context exception when attaching to a QEMU target running a cortex-a57 cpu. This seems very similar to #2153.

### Steps to reproduce

This happens for me when debugging the `elderly` challenge from this m0lecon finals CTF that runs a cortex-a57 Linux kernel
[elderly_kpwn.tar.gz](https://github.com/user-attachments/files/19402465/elderly_kpwn.tar.gz)

When attaching to the QEMU target with `aarch64-linux-gnu-gdb`, I have the following exception:

```
> $ aarch64-linux-gnu-gdb --ex "target remote :1234" --ex "c"                                                        
GNU gdb (GDB) 16.2
Copyright (C) 2024 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "--host=x86_64-pc-linux-gnu --target=aarch64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
pwndbg: loaded 187 pwndbg commands and 46 shell commands. Type pwndbg [--shell | --all] [filter] for a list.
pwndbg: created $rebase, $base, $hex2ptr, $argv, $envp, $argc, $environ, $bn_sym, $bn_var, $bn_eval, $ida GDB functions (can be used with print/break)
Remote debugging using :1234
warning: No executable has been specified and target does not support
determining executable automatically.  Try using the "file" command.
0xffffa429260f5678 in ?? ()
Continuing.
^C
Program received signal SIGINT, Interrupt.
0xffffa429260f5678 in ?? ()
------- tip of the day (disable with set show-tips off) -------
Use hi to see if a an address belongs to a glibc heap chunk
Exception occurred: context: address must be positive (<class 'AssertionError'>)
For more info invoke `set exception-verbose on` and rerun the command
or debug it by yourself with `set exception-debugger on`
pwndbg> set exception-verbose on
Set whether to print a full stacktrace for exceptions raised in Pwndbg commands to 'on'.
pwndbg> ctx
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /home/pazu/Tools/devel/pwndbg/pwndbg/commands/__init__.py:187 in __call__                        │
│                                                                                                  │
│   184 │                                                                                          │
│   185 │   def __call__(self, *args: Any, **kwargs: Any) -> str | None:                           │
│   186 │   │   try:                                                                               │
│ ❱ 187 │   │   │   return self.function(*args, **kwargs)                                          │
│   188 │   │   except TypeError:                                                                  │
│   189 │   │   │   print(f"{self.function.__name__.strip()!r}: {self.function.__doc__.strip()}"   │
│   190 │   │   │   pwndbg.exception.handle(self.function.__name__)                                │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/commands/context.py:728 in context                          │
│                                                                                                  │
│    725 │   │   │   │   result_settings[target].update(settings)                                  │
│    726 │   │   │   │   with target as out:                                                       │
│    727 │   │   │   │   │   result[target].extend(                                                │
│ ❱  728 │   │   │   │   │   │   func(                                                             │
│    729 │   │   │   │   │   │   │   target=out,                                                   │
│    730 │   │   │   │   │   │   │   width=settings.get("width", None),                            │
│    731 │   │   │   │   │   │   │   with_banner=settings.get("banner_top", True),                 │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/commands/context.py:390 in _serve_context_history           │
│                                                                                                  │
│    387 │   │   │   ):                                                                            │
│    388 │   │   │   │   current_output = context_history[section_name][-1]                        │
│    389 │   │   │   else:                                                                         │
│ ❱  390 │   │   │   │   current_output = function(*a, **kw)                                       │
│    391 │   │   │   if (                                                                          │
│    392 │   │   │   │   len(context_history[section_name]) == 0                                   │
│    393 │   │   │   │   or context_history[section_name][-1] != current_output                    │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/commands/context.py:984 in context_disasm                   │
│                                                                                                  │
│    981 │   if cs is not None and cs.syntax != syntax:                                            │
│    982 │   │   pwndbg.lib.cache.clear_caches()                                                   │
│    983 │                                                                                         │
│ ❱  984 │   result = try_emulate_if_bug_disable(                                                  │
│    985 │   │   lambda: pwndbg.aglib.nearpc.nearpc(                                               │
│    986 │   │   │   lines=disasm_lines // 2,                                                      │
│    987 │   │   │   emulate=bool(not pwndbg.config.emulate == "off"),                             │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/commands/context.py:960 in try_emulate_if_bug_disable       │
│                                                                                                  │
│    957                                                                                           │
│    958 def try_emulate_if_bug_disable(handler: Callable[[], T]) -> T:                            │
│    959 │   try:                                                                                  │
│ ❱  960 │   │   return handler()                                                                  │
│    961 │   except U.UcError as e:                                                                │
│    962 │   │   print(                                                                            │
│    963 │   │   │   message.warn(                                                                 │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/commands/context.py:985 in <lambda>                         │
│                                                                                                  │
│    982 │   │   pwndbg.lib.cache.clear_caches()                                                   │
│    983 │                                                                                         │
│    984 │   result = try_emulate_if_bug_disable(                                                  │
│ ❱  985 │   │   lambda: pwndbg.aglib.nearpc.nearpc(                                               │
│    986 │   │   │   lines=disasm_lines // 2,                                                      │
│    987 │   │   │   emulate=bool(not pwndbg.config.emulate == "off"),                             │
│    988 │   │   │   use_cache=True,                                                               │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/nearpc.py:143 in nearpc                               │
│                                                                                                  │
│   140 │   #         for line in symtab.linetable():                                              │
│   141 │   #             pc_to_linenos[line.pc].append(line.line)                                 │
│   142 │                                                                                          │
│ ❱ 143 │   instructions, index_of_pc = pwndbg.aglib.disasm.near(                                  │
│   144 │   │   pc, lines, emulate=emulate, show_prev_insns=not repeat, use_cache=use_cache, lin   │
│   145 │   )                                                                                      │
│   146                                                                                            │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/disasm/__init__.py:500 in near                        │
│                                                                                                  │
│   497 │   │   next_addresses_cache.add(target)                                                   │
│   498 │   │                                                                                      │
│   499 │   │   # The emulator is stepped within this call                                         │
│ ❱ 500 │   │   insn = one(target, emu, put_cache=True)                                            │
│   501 │   │                                                                                      │
│   502 │   │   if insn:                                                                           │
│   503 │   │   │   insns.append(insn)                                                             │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/disasm/__init__.py:245 in one                         │
│                                                                                                  │
│   242 │   │   return None                                                                        │
│   243 │                                                                                          │
│   244 │   # A for loop in case this returns an empty list                                        │
│ ❱ 245 │   for insn in get(address, 1, emu, enhance=enhance, from_cache=from_cache, put_cache=p   │
│   246 │   │   if put_backward_cache:                                                             │
│   247 │   │   │   backward_cache[insn.next] = insn.address                                       │
│   248 │   │   return insn                                                                        │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/disasm/__init__.py:280 in get                         │
│                                                                                                  │
│   277 │                                                                                          │
│   278 │   retval: List[PwndbgInstruction] = []                                                   │
│   279 │   for _ in range(instructions):                                                          │
│ ❱ 280 │   │   i = get_one_instruction(                                                           │
│   281 │   │   │   address, emu, enhance=enhance, from_cache=from_cache, put_cache=put_cache      │
│   282 │   │   )                                                                                  │
│   283 │   │   if i is None:                                                                      │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/disasm/__init__.py:218 in get_one_instruction         │
│                                                                                                  │
│   215 │   │   pwn_ins: PwndbgInstruction = PwndbgInstructionImpl(ins)                            │
│   216 │   │                                                                                      │
│   217 │   │   if enhance:                                                                        │
│ ❱ 218 │   │   │   pwndbg.aglib.disasm.arch.DisassemblyAssistant.enhance(pwn_ins, emu)            │
│   219 │   │                                                                                      │
│   220 │   │   if put_cache:                                                                      │
│   221 │   │   │   computed_instruction_cache[address] = pwn_ins                                  │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/disasm/arch.py:250 in enhance                         │
│                                                                                                  │
│    247 │   │   enhancer._enhance_conditional(instruction, emu)                                   │
│    248 │   │                                                                                     │
│    249 │   │   # Set the .target and .next fields                                                │
│ ❱  250 │   │   enhancer._enhance_next(instruction, emu, jump_emu)                                │
│    251 │   │                                                                                     │
│    252 │   │   if bool(pwndbg.config.disasm_annotations):                                        │
│    253 │   │   │   enhancer._set_annotation_string(instruction, emu)                             │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/disasm/arch.py:696 in _enhance_next                   │
│                                                                                                  │
│    693 │   │                                                                                     │
│    694 │   │   if instruction.has_jump_target:                                                   │
│    695 │   │   │   # Only bother doing the symbol lookup if this is a jump                       │
│ ❱  696 │   │   │   instruction.target_string = MemoryColor.get_address_or_symbol(instruction.ta  │
│    697 │   │                                                                                     │
│    698 │   │   if (                                                                              │
│    699 │   │   │   instruction.operands                                                          │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/color/memory.py:48 in get_address_or_symbol                 │
│                                                                                                  │
│    45 │   """                                                                                    │
│    46 │   Convert and colorize address to symbol if it can be resolved, else return colorized    │
│    47 │   """                                                                                    │
│ ❱  48 │   return attempt_colorized_symbol(address) or get(address)                               │
│    49                                                                                            │
│    50                                                                                            │
│    51 def attempt_colorized_symbol(address: int) -> str | None:                                  │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/color/memory.py:55 in attempt_colorized_symbol              │
│                                                                                                  │
│    52 │   """                                                                                    │
│    53 │   Convert address to colorized symbol (if symbol is there), else None                    │
│    54 │   """                                                                                    │
│ ❱  55 │   symbol = pwndbg.aglib.symbol.resolve_addr(address)                                     │
│    56 │   if symbol:                                                                             │
│    57 │   │   return get(address, symbol)                                                        │
│    58 │   else:                                                                                  │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/lib/cache.py:162 in decorator                               │
│                                                                                                  │
│   159 │   │   │   │   if cached_value is not _NOT_FOUND_IN_CACHE:                                │
│   160 │   │   │   │   │   return cached_value                                                    │
│   161 │   │   │   │                                                                              │
│ ❱ 162 │   │   │   │   value = func(*a, **kw)                                                     │
│   163 │   │   │   │                                                                              │
│   164 │   │   │   │   # Sanity check: its not perfect and won't cover all cases like ([],)       │
│   165 │   │   │   │   # but it should be good enough                                             │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/symbol.py:109 in resolve_addr                         │
│                                                                                                  │
│   106 │   Resolution is performed in the following order:                                        │
│   107 │   - Global scope symbols.                                                                │
│   108 │   """                                                                                    │
│ ❱ 109 │   assert addr >= 0, "address must be positive"                                           │
│   110 │                                                                                          │
│   111 │   symbol_name = pwndbg.dbg.selected_inferior().symbol_name_at_address(addr)              │
│   112 │   if symbol_name:                                                                        │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
AssertionError: address must be positive
If that is an issue, you can report it on https://github.com/pwndbg/pwndbg/issues
(Please don't forget to search if it hasn't been reported before)
To generate the report and open a browser, you may run `bugreport --run-browser`
PS: Pull requests are welcome

```

Session history:
```
set exception-verbose on
ctx
```

### My setup

```
Pwndbg:   2025.02.19 build: f9a37606
Python:   3.13.2 (main, Feb  5 2025, 08:05:21) [GCC 14.2.1 20250128]
GDB:      16.2
Capstone: 5.0.5
Unicorn:  2.1.3
Pwnlib:   4.14.0
OS: Arch Linux (Linux-6.13.7-hardened1-1-hardened-x86_64-with-glibc2.41, little endian)
OS ABI: #1 SMP PREEMPT_DYNAMIC Tue, 18 Mar 2025 08:18:37 +0000
Charset: utf-8
Terminal width: 118, height: 61
Target Arch: aarch64
ARM: arm abi:  The current ARM ABI is "auto".
arm apcs32:  Usage of ARM 32-bit mode is on.
arm disassembler:  The disassembly style is "".
arm fallback-mode:  The current execution mode assumed (when symbols are unavailable) is "auto".
arm force-mode:  The current execution mode assumed (even when symbols are available) is "auto".
arm fpu:  The current ARM floating point model is "auto".
arm unwind-secure-frames:  Usage of non-secure to secure exception stack unwinding is on.
```

I thought it was due to my environment, but I face the same issue on a Debian vm with gdb-multiarch. Like #2153, it seems that the wrong architecture is picked.
Also, I'm open to start working on a PR if we confirm there is something to fix :)

Thanks for catching this @patryk4815! I can confirm it works now for this target!

However, I spotted the same issue in another aarch64 kernel challenge (https://github.com/HeroCTF/HeroCTF_v6/tree/main/Pwn/Buafllet). Even with this temporary fix, I can still see the same exception sometimes (not for all instructions):
```
pwndbg> si
0xffff800080140da4 in ?? ()
LEGEND: STACK | HEAP | CODE | DATA | WX | RODATA
────────────────────────────────[ REGISTERS / show-flags off / show-compact-regs off ]────────────────────────────────
*X0   1
 X1   0xffff000007f78010 ◂— 0xf
 X2   1
 X3   0
 X4   0xffff800081cd3be0
 X5   0xffff000007f79bd0 ◂— 0x2d9c
 X6   0xffff800081cd3bd0
 X7   0xc0
 X8   0xffff000002d68e88 ◂— 0xa59db2bcfcdb8432
 X9   0xa59db2bcfcdb8432
 X10  0x919f11b7b3c00259
 X11  1
 X12  0
 X13  0xfa83b2da
 X14  0
 X15  0x12486ca0002
 X16  0
 X17  2
 X18  0xffffffffffffffff
 X19  0xffff800081cd2010
 X20  0xffff000007f78010 ◂— 0xf
 X21  0xffff7fff862a6000
 X22  0xffff000002d68000 ◂— 8
 X23  1
 X24  0
 X25  0xffff000002d68000 ◂— 8
 X26  0
 X27  0
 X28  0
 X29  0xffff800082993da0 —▸ 0xffff800082993e20 —▸ 0xffff800082993e70 —▸ 0xffff800082993e90 —▸ 0xffff800082993fe0 ◂— ...
 SP   0xffff800082993da0 —▸ 0xffff800082993e20 —▸ 0xffff800082993e70 —▸ 0xffff800082993e90 —▸ 0xffff800082993fe0 ◂— ...
 LR   0x80828000800d967c ◂— 0x80828000800d967c
*PC   0xffff800080140da4 ◂— bl #0xffff8000801409a0
────────────────────────────────────────[ DISASM / aarch64 / set emulate on ]─────────────────────────────────────────
   0xffff800080140c7c    ldr    w23, [x0, x21]            W23, [0xffff000007f72008] => 1
   0xffff800080140c80    ldr    x1, [x20, #0x90]          X1, [0xffff000007f780a0] => 0
   0xffff800080140c84  ✔ cbz    x1, #0xffff800080140d9c     <0xffff800080140d9c>
    ↓
   0xffff800080140d9c    mov    x1, x20                   X1 => 0xffff000007f78010 ◂— 0xf
   0xffff800080140da0    mov    w0, w23                   W0 => 1
 ► 0xffff800080140da4    bl     #0xffff8000801409a0         <0xffff8000801409a0>
 
   0xffff800080140da8    tbz    w0, #0, #0xffff800080140ca4
 
   0xffff800080140dac    mov    w1, w23
   0xffff800080140db0    mov    x0, x20
   0xffff800080140db4    bl     #0xffff800080140aa4
 
   0xffff800080140db8    mov    x22, x0
──────────────────────────────────────────────────────[ STACK ]───────────────────────────────────────────────────────
00:0000│ x29 sp 0xffff800082993da0 —▸ 0xffff800082993e20 —▸ 0xffff800082993e70 —▸ 0xffff800082993e90 —▸ 0xffff800082993fe0 ◂— ...
01:0008│        0xffff800082993da8 —▸ 0x80828000800d967c ◂— 0x80828000800d967c
02:0010│        0xffff800082993db0 ◂— 1
03:0018│        0xffff800082993db8 —▸ 0xffff800082409a40 ◂— 3
04:0020│        0xffff800082993dc0 —▸ 0xffff800082409b44 ◂— 0x100000000
05:0028│        0xffff800082993dc8 —▸ 0xffff000002d68000 ◂— 8
06:0030│        0xffff800082993dd0 ◂— 0
07:0038│        0xffff800082993dd8 ◂— 0
────────────────────────────────────────────────────[ BACKTRACE ]─────────────────────────────────────────────────────
 ► 0 0xffff800080140da4 None
   1 0xffff800082409a40 None
────────────────────────────────────────────────[ THREADS (2 TOTAL) ]─────────────────────────────────────────────────
  ► 2   "" stopped: 0xffff800080140da4
    1   "" stopped: 0xffff80008105d558
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
pwndbg> si
0xffff8000801409a0 in ?? ()
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /home/pazu/Tools/devel/pwndbg/pwndbg/commands/__init__.py:187 in __call__                        │
│                                                                                                  │
│   184 │                                                                                          │
│   185 │   def __call__(self, *args: Any, **kwargs: Any) -> str | None:                           │
│   186 │   │   try:                                                                               │
│ ❱ 187 │   │   │   return self.function(*args, **kwargs)                                          │
│   188 │   │   except TypeError:                                                                  │
│   189 │   │   │   print(f"{self.function.__name__.strip()!r}: {self.function.__doc__.strip()}"   │
│   190 │   │   │   pwndbg.exception.handle(self.function.__name__)                                │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/commands/context.py:728 in context                          │
│                                                                                                  │
│    725 │   │   │   │   result_settings[target].update(settings)                                  │
│    726 │   │   │   │   with target as out:                                                       │
│    727 │   │   │   │   │   result[target].extend(                                                │
│ ❱  728 │   │   │   │   │   │   func(                                                             │
│    729 │   │   │   │   │   │   │   target=out,                                                   │
│    730 │   │   │   │   │   │   │   width=settings.get("width", None),                            │
│    731 │   │   │   │   │   │   │   with_banner=settings.get("banner_top", True),                 │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/commands/context.py:390 in _serve_context_history           │
│                                                                                                  │
│    387 │   │   │   ):                                                                            │
│    388 │   │   │   │   current_output = context_history[section_name][-1]                        │
│    389 │   │   │   else:                                                                         │
│ ❱  390 │   │   │   │   current_output = function(*a, **kw)                                       │
│    391 │   │   │   if (                                                                          │
│    392 │   │   │   │   len(context_history[section_name]) == 0                                   │
│    393 │   │   │   │   or context_history[section_name][-1] != current_output                    │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/commands/context.py:984 in context_disasm                   │
│                                                                                                  │
│    981 │   if cs is not None and cs.syntax != syntax:                                            │
│    982 │   │   pwndbg.lib.cache.clear_caches()                                                   │
│    983 │                                                                                         │
│ ❱  984 │   result = try_emulate_if_bug_disable(                                                  │
│    985 │   │   lambda: pwndbg.aglib.nearpc.nearpc(                                               │
│    986 │   │   │   lines=disasm_lines // 2,                                                      │
│    987 │   │   │   emulate=bool(not pwndbg.config.emulate == "off"),                             │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/commands/context.py:960 in try_emulate_if_bug_disable       │
│                                                                                                  │
│    957                                                                                           │
│    958 def try_emulate_if_bug_disable(handler: Callable[[], T]) -> T:                            │
│    959 │   try:                                                                                  │
│ ❱  960 │   │   return handler()                                                                  │
│    961 │   except U.UcError as e:                                                                │
│    962 │   │   print(                                                                            │
│    963 │   │   │   message.warn(                                                                 │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/commands/context.py:985 in <lambda>                         │
│                                                                                                  │
│    982 │   │   pwndbg.lib.cache.clear_caches()                                                   │
│    983 │                                                                                         │
│    984 │   result = try_emulate_if_bug_disable(                                                  │
│ ❱  985 │   │   lambda: pwndbg.aglib.nearpc.nearpc(                                               │
│    986 │   │   │   lines=disasm_lines // 2,                                                      │
│    987 │   │   │   emulate=bool(not pwndbg.config.emulate == "off"),                             │
│    988 │   │   │   use_cache=True,                                                               │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/nearpc.py:143 in nearpc                               │
│                                                                                                  │
│   140 │   #         for line in symtab.linetable():                                              │
│   141 │   #             pc_to_linenos[line.pc].append(line.line)                                 │
│   142 │                                                                                          │
│ ❱ 143 │   instructions, index_of_pc = pwndbg.aglib.disasm.near(                                  │
│   144 │   │   pc, lines, emulate=emulate, show_prev_insns=not repeat, use_cache=use_cache, lin   │
│   145 │   )                                                                                      │
│   146                                                                                            │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/disasm/__init__.py:500 in near                        │
│                                                                                                  │
│   497 │   │   next_addresses_cache.add(target)                                                   │
│   498 │   │                                                                                      │
│   499 │   │   # The emulator is stepped within this call                                         │
│ ❱ 500 │   │   insn = one(target, emu, put_cache=True)                                            │
│   501 │   │                                                                                      │
│   502 │   │   if insn:                                                                           │
│   503 │   │   │   insns.append(insn)                                                             │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/disasm/__init__.py:245 in one                         │
│                                                                                                  │
│   242 │   │   return None                                                                        │
│   243 │                                                                                          │
│   244 │   # A for loop in case this returns an empty list                                        │
│ ❱ 245 │   for insn in get(address, 1, emu, enhance=enhance, from_cache=from_cache, put_cache=p   │
│   246 │   │   if put_backward_cache:                                                             │
│   247 │   │   │   backward_cache[insn.next] = insn.address                                       │
│   248 │   │   return insn                                                                        │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/disasm/__init__.py:280 in get                         │
│                                                                                                  │
│   277 │                                                                                          │
│   278 │   retval: List[PwndbgInstruction] = []                                                   │
│   279 │   for _ in range(instructions):                                                          │
│ ❱ 280 │   │   i = get_one_instruction(                                                           │
│   281 │   │   │   address, emu, enhance=enhance, from_cache=from_cache, put_cache=put_cache      │
│   282 │   │   )                                                                                  │
│   283 │   │   if i is None:                                                                      │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/disasm/__init__.py:218 in get_one_instruction         │
│                                                                                                  │
│   215 │   │   pwn_ins: PwndbgInstruction = PwndbgInstructionImpl(ins)                            │
│   216 │   │                                                                                      │
│   217 │   │   if enhance:                                                                        │
│ ❱ 218 │   │   │   pwndbg.aglib.disasm.arch.DisassemblyAssistant.enhance(pwn_ins, emu)            │
│   219 │   │                                                                                      │
│   220 │   │   if put_cache:                                                                      │
│   221 │   │   │   computed_instruction_cache[address] = pwn_ins                                  │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/disasm/arch.py:253 in enhance                         │
│                                                                                                  │
│    250 │   │   enhancer._enhance_next(instruction, emu, jump_emu)                                │
│    251 │   │                                                                                     │
│    252 │   │   if bool(pwndbg.config.disasm_annotations):                                        │
│ ❱  253 │   │   │   enhancer._set_annotation_string(instruction, emu)                             │
│    254 │   │                                                                                     │
│    255 │   │   # Disable emulation after CALL instructions. We do it after enhancement, as we c  │
│    256 │   │   # to determine the call's target address.                                         │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/disasm/aarch64.py:287 in _set_annotation_string       │
│                                                                                                  │
│   284 │   │   elif instruction.id in AARCH64_EMULATED_ANNOTATIONS:                               │
│   285 │   │   │   self._common_generic_register_destination(instruction, emu)                    │
│   286 │   │   else:                                                                              │
│ ❱ 287 │   │   │   self.annotation_handlers.get(instruction.id, lambda *a: None)(                 │
│   288 │   │   │   │   instruction, emu                                                           │
│   289 │   │   │   )                                                                              │
│   290                                                                                            │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/disasm/aarch64.py:300 in _handle_adrp                 │
│                                                                                                  │
│   297 │   │   │                                                                                  │
│   298 │   │   │   addresses = self._telescope(address, TELESCOPE_DEPTH, instruction, emu)        │
│   299 │   │   │                                                                                  │
│ ❱ 300 │   │   │   telescope = self._telescope_format_list(addresses, TELESCOPE_DEPTH, emu)       │
│   301 │   │   │                                                                                  │
│   302 │   │   │   instruction.annotation = register_assign(result_operand.str, telescope)        │
│   303                                                                                            │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/disasm/arch.py:543 in _telescope_format_list          │
│                                                                                                  │
│    540 │   │   enhance_string_len = int(pwndbg.config.disasm_telescope_string_length)            │
│    541 │   │                                                                                     │
│    542 │   │   if emu:                                                                           │
│ ❱  543 │   │   │   return emu.format_telescope_list(                                             │
│    544 │   │   │   │   addresses, limit, enhance_string_len=enhance_string_len                   │
│    545 │   │   │   )                                                                             │
│    546 │   │   else:                                                                             │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/emu/emulator.py:378 in format_telescope_list                │
│                                                                                                  │
│   375 │   │   # Colorize the chain                                                               │
│   376 │   │   rest = []                                                                          │
│   377 │   │   for link in chain:                                                                 │
│ ❱ 378 │   │   │   symbol = pwndbg.aglib.symbol.resolve_addr(link) or None                        │
│   379 │   │   │   if symbol:                                                                     │
│   380 │   │   │   │   symbol = f"{link:#x} ({symbol})"                                           │
│   381 │   │   │   rest.append(M.get(link, symbol))                                               │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/lib/cache.py:162 in decorator                               │
│                                                                                                  │
│   159 │   │   │   │   if cached_value is not _NOT_FOUND_IN_CACHE:                                │
│   160 │   │   │   │   │   return cached_value                                                    │
│   161 │   │   │   │                                                                              │
│ ❱ 162 │   │   │   │   value = func(*a, **kw)                                                     │
│   163 │   │   │   │                                                                              │
│   164 │   │   │   │   # Sanity check: its not perfect and won't cover all cases like ([],)       │
│   165 │   │   │   │   # but it should be good enough                                             │
│                                                                                                  │
│ /home/pazu/Tools/devel/pwndbg/pwndbg/aglib/symbol.py:109 in resolve_addr                         │
│                                                                                                  │
│   106 │   Resolution is performed in the following order:                                        │
│   107 │   - Global scope symbols.                                                                │
│   108 │   """                                                                                    │
│ ❱ 109 │   assert addr >= 0, "address must be positive"                                           │
│   110 │                                                                                          │
│   111 │   symbol_name = pwndbg.dbg.selected_inferior().symbol_name_at_address(addr)              │
│   112 │   if symbol_name:                                                                        │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
AssertionError: address must be positive
```

--- a/pwndbg/aglib/disasm/aarch64.py
+++ b/pwndbg/aglib/disasm/aarch64.py
@@ -281,7 +281,7 @@ class DisassemblyAssistant(pwndbg.aglib.disasm.arch.DisassemblyAssistant):
     def _handle_adrp(self, instruction: PwndbgInstruction, emu: Emulator) -> None:
         result_operand, right = instruction.operands
         if result_operand.str and right.before_value is not None:
-            address = right.before_value
+            address = right.before_value & pwndbg.aglib.arch.ptrmask
 
             TELESCOPE_DEPTH = max(0, int(pwndbg.config.disasm_telescope_depth))
 
@@ -348,7 +348,9 @@ class DisassemblyAssistant(pwndbg.aglib.disasm.arch.DisassemblyAssistant):
 
         if len(instruction.operands) > 0:
             # For all AArch64 branches, the target is either an immediate or a register and is the last operand
-            return instruction.operands[-1].before_value
+            if (val := instruction.operands[-1].before_value) is not None:
+                return val & pwndbg.aglib.arch.ptrmask
+            return None
         elif instruction.id == ARM64_INS_RET:
             # If this is a ret WITHOUT an operand, it means we should read from the LR/x30 register
             return super()._read_register_name(instruction, "lr", emu)

--- a/pwndbg/aglib/disasm/arch.py
+++ b/pwndbg/aglib/disasm/arch.py
@@ -691,7 +691,7 @@ class DisassemblyAssistant:
         if instruction.target is None:
             instruction.target = instruction.next
 
-        if instruction.has_jump_target:
+        if instruction.has_jump_target and instruction.target >= 0:
             # Only bother doing the symbol lookup if this is a jump
             instruction.target_string = MemoryColor.get_address_or_symbol(instruction.target)
 

--- a/pwndbg/emu/emulator.py
+++ b/pwndbg/emu/emulator.py
@@ -190,6 +190,7 @@ BANNED_INSTRUCTIONS = {
     "mips": {C.mips.MIPS_INS_RDHWR},
     "arm": ARM_BANNED_INSTRUCTIONS,
     "armcm": ARM_BANNED_INSTRUCTIONS,
+    "aarch64": {C.arm64.ARM64_INS_MRS},
 }
 
 # https://github.com/unicorn-engine/unicorn/issues/550

